#11 공통 컴포넌트 (카드 프로필)

gpfksdlrn opened the pull request:
### 📌 관련 이슈
- #11 

### 📋 작업 내용
- [x] 카드 프로필 이미지 공통 컴포넌트
- [x] 카드 기본카드형
- [x] 카드 후원카드형

### 📷 결과 및 스크린샷
|제목|내용|
|-----|----|
|기본카드형|<img width="199" alt="스크린샷 2025-04-18 오후 11 21 04" src="https://github.com/user-attachments/assets/41707f05-9cbd-448b-afa1-bfde41bfa273" />|
|후원카드형<br>(위의 이미지: 'm'사이즈, <br>아래의 이미지: 's'사이즈)|<img width="200" alt="image" src="https://github.com/user-attachments/assets/3a3ce393-fcf5-4d54-9a3b-0115a90bc8cc" />|

### 🔎 참고 (선택)
버튼은 임시로 구현한 것으로, 실제 컴포넌트 디자인과 다를 수 있는 점 참고 부탁드립니다.

gpfksdlrn committed:
#11 feat: IdolCardList에 size props 및 context 추가, 카드 사이즈별 스타일 적용

gpfksdlrn committed:
#11 feat: 후원 정보 컴포넌트

gpfksdlrn committed:
#11 feat: 카드 프로필 이미지 컴포넌트

gpfksdlrn committed:
#11 fix: import문 누락

gpfksdlrn committed:
#11 feat: 후원 정보 Footer 영역 컴포넌트

gpfksdlrn committed:
#11 docs: CardImg 및 IdolCardList 컴포넌트에 JSDoc 주석 추가

gpfksdlrn committed:
#11 feat: isHover일 때 노출되는 '후원하기' 버튼 UI(추후 버튼 컴포넌트로 변경 필요)

MyungJiwoo commented on src/components/card/IdolCard.jsx:
```diff
@@ -0,0 +1,193 @@
+  return (
+    <>
+      <div className='caption-text flex items-center justify-between pb-4'>
+        <span className='text-[var(--color-brand-1)]'>₩ {credit}</span>
```
₩ 부분 대신에 크레딧 아이콘을 넣어주시면 서비스의 특징이 더 잘 드러날 것 같습니다!

gpfksdlrn replied:
네! 수정하도록 하겠습니다.

MyungJiwoo commented on src/components/common/CardImg.jsx:
```diff
@@ -0,0 +1,34 @@
+      <img
+        src={src}
+        alt={title}
+        className='w-full rounded-2xl object-contain'
```
<img width="412" alt="image" src="https://github.com/user-attachments/assets/9168118d-71d3-4af3-8f51-c8f3445e5393" />

현재 이미지의 높이가 설정되어 있지 않아 사진마다 컴포넌트의 크기가 달라지는 것 같습니다..ㅠㅠ
`목록 페이지 - 후원을 기다리는 조공`에서는 정사각형 비율이고, `후원하기 모달창`에서는 다른 비율인 점 참고해주시면 좋을 것 같아요!
시간상 빠듯하시다면 개인적으로는 같은 비율로 통일하셔도 괜찮을 것 같습니다~

비율에 관한 스타일링은 직접 설정해주셔도 되고, 아래 링크처럼 `Aspect Ratio`도 추천드립니다!
https://velog.io/@springhana/Tailwind-CSS%EC%9D%98-Ascpect-Ratio%EB%A1%9C-%EC%82%AC%EC%9D%B4%EC%A6%88-%EB%A7%9F%EC%B6%94%EA%B8%B0

gpfksdlrn replied:
말씀해주신 대로 현재 이미지 높이가 고정되어 있지 않아서 카드 크기 차이가 발생했던 것 같아요.  

시간이 허락된다면 목록/모달 별로 적절한 비율을 맞추고,  
그게 어려울 경우 말씀해주신 것처럼 우선은 통일된 비율로 적용해보는 방향으로 진행할게요!
추천해주신 `aspect-ratio` 활용 방안도 검토해보겠습니다!

좋은 자료까지 함께 공유해주셔서 감사합니다!

gpfksdlrn committed:
#11 style: 기본카드형 가로 크기 변경

MyungJiwoo commented on src/components/card/IdolCard.jsx:
```diff
@@ -0,0 +1,193 @@
+        isHover={isHover}
+      />
+      {isHover && (
+        <button
+          onClick={onClick}
+          className='absolute bottom-[1rem] left-[50%] z-10 -translate-x-1/2 cursor-pointer bg-gradient-to-t from-[#F96D69] to-[#FF9D95] px-6 py-2 text-sm font-semibold text-white'
+        >
+          후원하기
+        </button>
```
현재 `isHover`에 따라 `후원하기 버튼`이 고정적으로 나오게 설정되어 있습니다. 
이때 버튼을 외부에서 받아올 수 있다면 재사용성이 더 높아질 것 같아요!

예를 들어, 
1. 외부에서 버튼 컴포넌트 **자체**를 받아옵니다.
```jsx
const IdolCardContext = createContext({
  id: null,
  src: '',
  location: '',
  title: '',
  credit: 0,
  daysLeft: 0,
  onClick: () => {},
  isHover: false,
  size: 'm',
  button: () => null,
});
```
저희는 재사용 버튼 컴포넌트가 있기 때문에 `() => null`로 초기화했습니다.
만약 jsx 요소를 전달한다면 `null`로 하시면 됩니다.

이때 버튼 컴포넌트 자체라는 것은 아래와 같은 형태입니다.
```jsx
const Button = () => {
  return (
    <button className='absolute bottom-[1rem] left-[50%] z-10 -translate-x-1/2 cursor-pointer bg-gradient-to-t from-[#F96D69] to-[#FF9D95] px-6 py-2 text-sm font-semibold text-white'>
      후원하기
    </button>
  );
};
```
<br />

2. 그리고 물론 context에 새로운 button이 생겼기 때문에 props에도 추가해줘야 합니다!
```jsx
const IdolCardList = ({ id, ... button, children }) => {
  const contextValue = { id, ... button};
...
```

<br />

3. 이제 이미지 컴포넌트 위에 외부에서 prop으로 받아온 버튼 컴포넌트를 띄웁니다.
```jsx
const IdolCardImg = () => {
  const { src, ... button: Button } = useContext(IdolCardContext);

  return (
    <div className='relative w-full'>
      <CardImg
        className='rounded-2xl object-contain'
        src={src}
        alt={title}
        isHover={isHover}
      />
      {Button && <Button />}
    </div>
  );
};
```
컴포넌트의 이름은 대문자로 시작하기 때문에 `button: Button`으로 이름을 변경합니다.
그리고 버튼 컴포넌트를 띄워줍니다. 이때 주의할 점은 버튼 컴포넌트가 있을 때만 보여줘야 하기 때문에 `Button && <Button />`과 같은 조건부가 들어가야 합니다.
또한, 버튼이 이미지 위에 떠야하기 때문에 `absolute` 속성이 있어야 합니다.

이렇게 사용하시면 외부에서 넘겨주는 버튼을 모두 사용할 수 있기 때문에 컴포넌트의 확장성을 높일 수 있습니다. 
하지만 이 방식은 구조가 조금 더 복잡해지고, 버튼 재사용 컴포넌트에 absolute 속성이 필요하다는 단점이 있을 것 같아요.

추가로, isHover 대신에 외부로 받아오는 Button 자체를 하나의 합성할 수 있는 자식 컴포넌트로 만들어두고, 이거에 따라 배경 그라디언트 레이어가 전달되도록 해도 좋을 것 같습니다. (`CardImg`에서 `{children}`으로 띄울 수 있게끔)
(이 부분은 저도 계속 고민해봤는데 잠깐 고민하기엔 구조가 좀 어렵네요....🥲)

gpfksdlrn replied:
정성스러운 피드백 감사합니다! 
예시까지 들어주셔서 이해하는데 정말 많은 도움이 되었습니다.

말씀해주신 것처럼 버튼을 외부에서 전달받아 렌더링하는 방식은 확실히 컴포넌트 재사용성과 확장성 측면에서 훨씬 유연할 것 같습니다.

초기엔 간단한 구현을 목표로 `isHover` 조건만으로 버튼을 고정 렌더링했는데, 말씀해주신 방식대로 context에 버튼을 넘기고, 필요한 위치에 조건부 렌더링하면 다양한 버튼 컴포넌트를 조합할 수 있어 구조적으로 더 깔끔할 것 같아요!

또, `CardImg`에서 `children`을 받아 합성하는 방식도 말씀하신 대로 확장성에 큰 도움이 될 수 있을 것 같아서 이 부분도 한번 고민해보겠습니다!

MyungJiwoo reviewed (changes requested):
합성 컴포넌트가 처음엔 엄청 복잡하고 어려우셨을텐데도 잘 구현해주신 것 같아요! 자세한 내용은 리뷰에 남겨두었지만 IdolCard 컴포넌트의 재사용성을 높이는 방법에 대해 더 고민해볼 수 있을 것 같습니다. (가볍게 생각하기엔 너무 어렵네요..ㅠㅠ 일단 저도 계속 고민해볼게요!) 
그리고 길게 남겨둔 부분은 참고해주시면 좋을 내용이고, 꼭 반영해주실 필요는 없습니다! 대신 아이콘이나 이미지 크기 부분은 수정 부탁드립니다. 🥲
고생 정말 많으셨습니다! 👍🏻👍🏻

gpfksdlrn committed:
#11 style: ₩ 기호를 Credit 아이콘으로 변경

gpfksdlrn committed:
#11 fix: 이미지 round 적용 및 정사각형 비율로 스타일 수정(aspect-ratio 적용)

gpfksdlrn committed:
#11 style: UI 스타일 정리 (className 정리, padding 조정, 배경색 제거)

gpfksdlrn committed:
#11 refactor: button prop 추가 및 context에 전달되도록 구조 확장
- 버튼 컴포넌트를 props로 받아 이미지 위에 오버레이로 렌더링 가능하도록 구조 확장
- button 기본값 null로 처리하여 선택적 사용 가능
- CardImg 내 children 렌더링 구조 유지
- 주석 및 예제 업데이트로 사용법 명확화

gpfksdlrn committed:
#11 refactor: CardImg 컴포넌트에서 children 있을 때만 ::before 오버레이 렌더링

MyungJiwoo commented on package.json:
```diff
@@ -23,6 +23,7 @@
   },
   "devDependencies": {
     "@eslint/js": "^9.21.0",
+    "@tailwindcss/aspect-ratio": "^0.4.2",
```
엇! tailwindcss 3.0 이상부터는 aspect-ratio가 기본 내장되었다는 얘기가 있네요!
https://velog.io/@bangina/Tailwind-CSS-V3.0

따로 설치하지 않아도 사용할 수 있을 것 같습니다!~

MyungJiwoo commented on src/components/common/CardImg.jsx:
```diff
@@ -22,15 +24,16 @@
 const CardImg = ({ src, title, children }) => {
+  const wrapperClassName = cn(
+    'relative aspect-[1/1] w-full overflow-hidden rounded-2xl',
+    children &&
+      "before:content-[''] before:absolute before:top-0 before:left-0 before:w-full before:h-full before:z-[3] before:bg-[linear-gradient(0deg,_rgb(2,0,14)_0%,_rgba(2,0,14,0)_80%)]",
```
오..!! 뒷 배경을 레이어로 컴포넌트화 해야 한다고 생각했었는데 CSS 가상 선택자를 활용하는 방법이 있었네요!
좋은 방법인 것 같아요. 배워갑니다 ..👍🏻

MyungJiwoo reviewed (approved):
컴포넌트를 확장하는 리팩토링이 정말 어렵다고 생각하는데 코드를 보면서 고민의 흔적이 느껴졌어요. 그럼에도 재사용 가능한 구조로 정리해주셔서 인상 깊었습니다. 정말 고생 많으셨어요! 👍🏻👍🏻

evaain706 commented on src/components/common/CardImg.jsx:
```diff
@@ -0,0 +1,41 @@
+  const wrapperClassName = cn(
+    'relative aspect-[1/1] w-full overflow-hidden rounded-2xl',
+    children &&
+      "before:content-[''] before:absolute before:top-0 before:left-0 before:w-full before:h-full before:z-[3] before:bg-[linear-gradient(0deg,_rgb(2,0,14)_0%,_rgba(2,0,14,0)_80%)]",
```
컴포넌트 구현 수고많으셨습니다! before를 통해 children존재시 그라데이션을 넣어주는 방식 잘 배워갑니다.

Yongmin0423 reviewed (approved):
고생 많으셨습니다!  합성 컴포넌트 어려운데 잘 구현해주셨네요! 캐러셀에서 카드 한 번 사용해보고 싶네요!

gpfksdlrn committed:
#11 chore: @tailwindcss/aspect-ratio 패키지 제거
Tailwind CSS 3.0부터 aspect-ratio 기능이 기본 내장되어 별도 플러그인 불필요
